src/store/utils/dereference-pipeline.ts

```diff
@@ -0,0 +1,62 @@
+import { Pipeline } from '@/lib/steps';
+
+export type PipelinesScopeContext = {
+  [pipelineName: string]: Pipeline;
+};
+
+/**
+ * Return a pipeline for the corresponding domain
+ */
+function _getPipelineForDomain(reference: string, pipelines: PipelinesScopeContext): Pipeline {
+  if (Object.keys(pipelines).includes(reference)) {
+    // eslint-disable-next-line @typescript-eslint/no-use-before-define
+    return dereferencePipelines(pipelines[reference], pipelines);
+  } else {
+    return [{ domain: reference, name: 'domain' }];
+  }
+}
+
+/**
+ * Dereference pipelines names in the current pipeline being edited, i.e.
+ * replaces references to pipelines or domains (by their names) to their corresponding
+ * pipelines
+ *
+ * @param pipeline the pipeline to translate and execute on the backend
+ * @param pipelines the pipelines stored in the Vuex store of the app, as an
+ * object with the pipeline name as key and the correspondinng pipeline as value
+ *
+ * @return the dereferenced pipeline
+ */
+export function dereferencePipelines(
+  pipeline: Pipeline,
+  pipelines: PipelinesScopeContext,
+): Pipeline {
+  const dereferencedPipeline: Pipeline = [];
+  for (const step of pipeline) {
+    switch (step.name) {
+      case 'domain':
+        // if domain is a pipeline replace step by selected dereferenced pipeline steps
+        dereferencedPipeline.push(..._getPipelineForDomain(step.domain as string, pipelines));
+        break;
+      case 'append':
+        dereferencedPipeline.push({
+          ...step,
+          pipelines: (step.pipelines as string[]).map(reference =>
```

Comment:
to avoid doing `step.pipelines as string[]` and have better autocomplete in this switch/case statement, you could try:
```
case 'append':
  const append_step = step as AppendStep;
  // then, use append_step.pipelines
  dereferencedPipeline.push({...append_step, pipelines: append_step.pipelines.map( ref => ...)
  break;
case (...):
```

Reply:
Actually, even this should be unnecessary, as Step should be a discriminated union](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) by its `name` property.
I wonder why this is not the case today

Reply:
@davinov , as now domain can be a Pipeline or a String, in any case, i need to force type otherwise, it won't work. I checked youre idea but it didn't worked neither :cry: !

Reply:
@alice-sevin ahhh I think I see what you mean now!
I think the culprit is the AppendStep type: there should be two.
The `AppendStep` and the `DereferencedAppendStep`
The `AppendStep` has `pipelines: string[] ` while the `DereferencedAppendStep` has `pipelines: Pipeline[]`

Also, differentiating the two in terms of types will make sure we know at some point which one we have (is it dereferenced, is it not?). Do you think that would make the code clearer and cleaner?

Reply:
Yeah i will try it, may be a very good way to help